# SVD da matriz de classificação de filmes  

Temos a matriz \(M\) (4 usuários × 5 filmes) com notas de 1 a 5. Zeros significam “ainda não avaliado”.

$$
M=\begin{bmatrix}
5 & 4 & 0 & 0 & 3\\
3 & 0 & 4 & 2 & 1\\
4 & 3 & 5 & 5 & 0\\
0 & 2 & 1 & 3 & 4
\end{bmatrix}
$$

A ideia é quebrar $M$ em $M = U\,\Sigma\,V^{\mathsf T}$ usando **SVD** para achar fatores latentes e prever notas faltantes.


In [ ]:
import numpy as np

M = np.array([
    [5, 4, 0, 0, 3],
    [3, 0, 4, 2, 1],
    [4, 3, 5, 5, 0],
    [0, 2, 1, 3, 4]
], dtype=float)

U, S, Vt = np.linalg.svd(M, full_matrices=False)

# Sigma completa (4×5): Primeiro, criamos uma matriz nula 4×5
# e depois colocamos os valores singulares (recebidos da função np.linalg.svd) na diagonal principal.
Sigma = np.zeros((4, 5))
np.fill_diagonal(Sigma, S)

# U (4×4): U é uma matriz ortonormal cujas colunas são os vetores singulares à esquerda.
# Vt (4×5): Vt é a transposta da matriz V, cujas linhas são os vetores singulares à direita.
np.set_printoptions(precision=3, suppress=True)
print("U (4×4):\n", U, "\n")
print("Σ (4×5):\n", Sigma, "\n")
print("Vᵀ (4×5):\n", Vt)


U (4×4):
 [[ 0.452  0.808 -0.378 -0.01 ]
 [ 0.42  -0.312 -0.141 -0.841]
 [ 0.715 -0.439 -0.097  0.536]
 [ 0.329  0.24   0.91  -0.077]] 

Σ (4×5):
 [[11.483  0.     0.     0.     0.   ]
 [ 0.     5.625  0.     0.     0.   ]
 [ 0.     0.     4.094  0.     0.   ]
 [ 0.     0.     0.     2.177  0.   ]] 

Vᵀ (4×5):
 [[ 0.556  0.402  0.486  0.47   0.269]
 [ 0.24   0.426 -0.569 -0.372  0.546]
 [-0.66   0.004 -0.034  0.48   0.578]
 [-0.198  0.648 -0.349  0.352 -0.543]]


### Matrizes resultantes (aproximados a 3 casas decimais)

$$
U \;\approx\;
\begin{bmatrix}
0.452 & 0.808 & -0.378 & -0.010\\
0.420 & -0.312 & -0.141 & -0.841\\
0.715 & -0.439 & -0.097 & 0.536\\
0.329 & 0.240 & 0.910 & -0.077
\end{bmatrix}
$$

$$
\Sigma \;\approx\;
\begin{bmatrix}
11.483 & 0      & 0     & 0     & 0\\
0      & 5.625 & 0     & 0     & 0\\
0      & 0      & 4.094 & 0     & 0\\
0      & 0      & 0     & 2.177 & 0
\end{bmatrix}
$$

$$
V^{\mathsf T} \;\approx\;
\begin{bmatrix}
\;0.556 & 0.402 & 0.486 & 0.470 & 0.269\\
\;0.240 & 0.426 & -0.569 & -0.372 & 0.546\\
-0.660 & 0.004 & -0.034 & 0.480 & 0.578\\
-0.198 & 0.648 & -0.349 & 0.352 & -0.543
\end{bmatrix}
$$

**Resumo rápido:**

* Os **valores singulares** em $\Sigma$ mostram que os **dois primeiros fatores** (11.483 e 5.625) carregam quase toda a informação.  
* Podemos truncar para $k=2$ e reconstruir $M_k$ para **estimar** as notas que faltam.  
